gke-deploy/core/gcs/gcs.go

```diff
@@ -0,0 +1,109 @@
+/*
+Copyright 2019 Google, Inc. All rights reserved.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+// Package gcs contains logic related to Google Cloud Storage.
+package gcs
+
+import (
+	"context"
+	"errors"
+	"log"
+	"time"
+
+	"github.com/GoogleCloudPlatform/cloud-builders/gke-deploy/services"
+)
+
+const TimeoutErr = "GCS timeout"
+
+var (
+	defaultTimeout = 1 * time.Hour
+	errGCSTimeout  = errors.New(TimeoutErr)
+)
+
+type GCS struct {
+	Timeout    time.Duration
+	Retries    int
+	Delay      time.Duration
+	GcsService services.GcsService
+}
+
+// Download copies file(s) from GCS. <dst> should be a directory, not a path to a file
+func (s *GCS) Download(ctx context.Context, src, dst string, recursive bool) error {
+	log.Printf("Downlaoding file(s) from GCS. Source: %s  Destination: %s.\n", src, dst)
+	return s.copyWithRetry(ctx, src, dst, recursive)
+}
+
+// Upload copies file(s) to GCS.
+func (s *GCS) Upload(ctx context.Context, src, dst string) error {
+	log.Printf("Uploading file(s) to GCS.  Source: %s  Destination: %s.\n", src, dst)
+	return s.copyWithRetry(ctx, src, dst, false)
+
+}
+
+// copyWithRetry is responsible for trying (and retrying) to call copyWithTimeout()
+// with appropriate retry backoff.
+func (s *GCS) copyWithRetry(ctx context.Context, src, dst string, recursive bool) error {
+	var err error
+	delay := s.Delay
+	for retryNum := 0; retryNum <= s.Retries; retryNum++ {
+
+		if retryNum > 0 {
+			time.Sleep(delay)
+			delay *= 2
```

Comment:
It seems to me like just a fixed delay would be sufficient here, instead of this exponential backoff since most of the time I would guess network issues are the cause of any temporary errors here. Also worth noting that a time.Duration is just an int64 under the hood so the default is 0 and this will infinitely loop. https://golang.org/pkg/time/#Duration

Reply:
I made the change to use a fixed delay. I don't quite get the infinite loop part though.